sycl/include/sycl/ext/oneapi/reduction.hpp

```diff
@@ -1684,13 +1991,164 @@ constexpr auto filterSequence(FunctorT F, std::index_sequence<Is...> Indices) {
   return filterSequenceHelper<T...>(F, Indices);
 }
 
+struct IsScalarReduction {
+  template <typename Reduction> struct Func {
+    static constexpr bool value =
+        (Reduction::dims == 0 && Reduction::num_elements == 1);
+  };
+};
+
+struct IsArrayReduction {
+  template <typename Reduction> struct Func {
+    static constexpr bool value =
+        (Reduction::dims == 1 && Reduction::num_elements > 1);
+  };
+};
+
+/// All scalar reductions are processed together; there is one loop of log2(N)
+/// steps, and each reduction uses its own storage.
+template <bool Pow2WG, bool IsOneWG, typename... Reductions, int Dims,
+          typename... LocalAccT, typename... OutAccT, typename... ReducerT,
+          typename... Ts, typename... BOPsT, size_t... Is>
+void reduCGFuncImplScalar(
+    nd_item<Dims> NDIt, ReduTupleT<LocalAccT...> LocalAccsTuple,
+    ReduTupleT<OutAccT...> OutAccsTuple, std::tuple<ReducerT...> &ReducersTuple,
+    ReduTupleT<Ts...> IdentitiesTuple, ReduTupleT<BOPsT...> BOPsTuple,
+    std::array<bool, sizeof...(Reductions)> InitToIdentityProps,
+    std::index_sequence<Is...> ReduIndices) {
+  size_t WGSize = NDIt.get_local_range().size();
+  size_t LID = NDIt.get_local_linear_id();
+  initReduLocalAccs<Pow2WG>(LID, WGSize, LocalAccsTuple, ReducersTuple,
+                            IdentitiesTuple, ReduIndices);
+  NDIt.barrier();
+
+  size_t PrevStep = WGSize;
+  for (size_t CurStep = PrevStep >> 1; CurStep > 0; CurStep >>= 1) {
+    if (LID < CurStep) {
+      // LocalReds[LID] = BOp(LocalReds[LID], LocalReds[LID + CurStep]);
+      reduceReduLocalAccs(LID, LID + CurStep, LocalAccsTuple, BOPsTuple,
+                          ReduIndices);
+    } else if (!Pow2WG && LID == CurStep && (PrevStep & 0x1)) {
+      // LocalReds[WGSize] = BOp(LocalReds[WGSize], LocalReds[PrevStep - 1]);
+      reduceReduLocalAccs(WGSize, PrevStep - 1, LocalAccsTuple, BOPsTuple,
+                          ReduIndices);
+    }
+    NDIt.barrier();
+    PrevStep = CurStep;
+  }
+
+  // Compute the partial sum/reduction for the work-group.
+  if (LID == 0) {
+    size_t GrID = NDIt.get_group_linear_id();
+    writeReduSumsToOutAccs<Pow2WG, IsOneWG>(
+        GrID, WGSize, (std::tuple<Reductions...> *)nullptr, OutAccsTuple,
```

Comment:
Is the cast required?

Reply:
I didn't write this code (I just moved it into `reduCGFuncImplScalar`) but I think the cast is required unless we change the definition of `writeReduSumsToOutAccs`.

My understanding is that this is basically a dummy argument that's being used  to help the compiler deduce the `Reductions...` template of `writeReduSumsToOutAccs`.  I think it should be possible to remove this pointer argument completely by passing `Reductions...` as an explicit template argument, though?  Something like:

```c++
    writeReduSumsToOutAccs<Pow2WG, IsOneWG, Reductions...>(
        GrID, WGSize, OutAccsTuple,
        LocalAccsTuple, BOPsTuple, IdentitiesTuple, InitToIdentityProps,
        ReduIndices);
```

Should I make that change?

Reply:
That code/pointer was needed to help deducer. 
The fix is already huge and extremely complex. Let's keep optimization/improvement of such places for separate PRs.

Reply:
> BTW, this reduction implementation is getting more and more classes. They are very convenient and help to avoid code-duplication, but the code becomes truly complex and difficult to move further without having some diagram of classes and 1-2 line description explaining which class is needed for what. It would be great to have such diagram eventually (maybe something similar to this https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/accessor.hpp#L92, but a bit more complex that that).

I completely agree.

There are a few other things we could to do simplify the code before we draw those diagrams, which might be worth it... We've already discussed how removing the legacy `accessor` interface would simplify things; I also saw some comments in the code suggesting that eventually we could remove everything except the parameter pack implementation, which might allow us to remove a few more specializations (or, at least, move them somewhere else).